### PR TITLE
ENH: Add array_like function to simplify input checking

### PR DESCRIPTION
**Pull request by bashtage**
Add a lightweight array subclass to simplify input validation and
standardization

- [X] tests added / passed. 
- [X] code/documentation is well formatted.  
- [X] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message).

### CODE REVIEW

**Comment by bashtage** [comment]
Input validation is a real pain point for me, especially when writing light-weight user-focused functions that don't pass through base.Model.  This is the start of some validation classes that can be used as 1 liners to provide users with good information about requires shape. ndim, etc.  This method also allows other inputs types (beyond ndarray and pandas) to be added across the entire code base wherever it is used.

**Comment by bashtage** [comment]
Probably closes #5925

**Comment by jbrockmendel** [comment]
If it’s just/mostly for validation, what do we get from making a subclass instead of a function?  My understanding is ndarray subclass support can be fragile

**Comment by bashtage** [comment]
I wrote a function version which is perfectly fine for unifying checks and conversion.  It didn't solve issues with handling pandas vs numpy and so I went for a very thin subclass which provides some pandas-focused services (1 for now).  This is so lightweight that I don't think the issues with subclasses manifest here.  For example, this subclass never propagates by construction (since it doesn't make sense).

**Comment by bashtage** [comment]
You are probably right that it isn't worth the effort since I have found some of the gotchas that aren't really documented AFAICT.  Probably safer to use a function.

File: statsmodels/tsa/arima_model.py [hunk 1]
```diff
@@ -421,9 +423,7 @@ def __init__(self, endog, order, exog=None, dates=None, freq=None,
                  missing='none'):
         super(ARMA, self).__init__(endog, exog, dates, freq, missing=missing)
         # GH 2575
-        _endog = endog if hasattr(endog, 'ndim') else np.asarray(endog)
-        if (_endog.ndim == 2 and _endog.shape[1] != 1) or _endog.ndim > 2:
-            raise ValueError('endog must be 1-d or 2-d with 1 column')
+        array_like(endog, 'endog')
```
Comment (jbrockmendel) [inline statsmodels/tsa/arima_model.py]: should this be `endog = array_like(endog, 'endog')`?

File: statsmodels/tsa/arima_model.py [hunk 2]
```diff
@@ -1535,7 +1534,7 @@ def forecast(self, steps=1, exog=None, alpha=.05):
         if exog is not None:
             #TODO: make a convenience function for this. we're using the
             # pattern elsewhere in the codebase
-            exog = np.asarray(exog)
+            exog = array_like(exog, 'exog', maxdim=2)
```
Comment (jbrockmendel) [inline statsmodels/tsa/arima_model.py]: would it make sense to handle the TODO above along with this validation?

File: statsmodels/tsa/arima_model.py (hunk 1 above)
Comment (bashtage) [reply statsmodels/tsa/arima_model.py]: This is just a check. endog isn't used below

Comment (bashtage) [reply statsmodels/tsa/arima_model.py]: All of the data handling is done somewhere in base.

**Pull request merged**